### PR TITLE
[azure][feat] Implement BaseResources for all Azure compute resources

### PR DESCRIPTION
**Pull request by 1101-1**
# Description

I have implemented several base resource classes, including `BaseInstance`, `BaseVolume`, `BaseInstanceType`, `BaseSnapshot`, and `BaseAutoScalingGroup`.

In the class mapping for `AzureVirtualMachine `(which inherits from `BaseInstance`), I've added fields like `instance_cores` and `instance_type`, which we can extract from Azure JSON. However, the fields `instance_memory` and `instance_status` are inaccessible because this information is not provided in the response.

For `AzureDisk` (based on` BaseVolume`), I've incorporated fields such as `volume_size`, `volume_type`, `volume_status`, `volume_iops`, `volume_throughput`, and `volume_encrypted`. Unfortunately, the `snapshot_before_delete` field from the base class remains unreachable because Azure does not supply this information.

In the case of `AzureVirtualMachineSize` (a subclass of `BaseInstanceType`), I've introduced fields for `instance_cores` and `instance_memory`. However, `instance_type`, `ondemand_cost`, and reservations remain empty since Azure does not provide the necessary data to populate these fields.

Regarding `AzureSnapshot` (derived from `BaseSnapshot`), I've included fields like `snapshot_status`, `volume_id`, `volume_size`, `encrypted`, and `owner_id`. Nevertheless, I am unable to retrieve information for the `owner_alias` and `description` fields since Azure does not provide this data.

Lastly, in the class mapping for `AzureVirtualMachineScaleSet` (based on` BaseAutoScalingGroup`), I have not added fields for `min_size` and `max_size` since Azure does not furnish this information.

Also, I've implemented a dictionary in my code called` VolumeStatusMapping` that serves as a mapping between string-based volume statuses and their corresponding `VolumeStatus` enum types like that:
```python
VolumeStatusMapping = {
    "ActiveSAS": VolumeStatus.IN_USE,
    "ActiveSASFrozen": VolumeStatus.BUSY,
    "ActiveUpload": VolumeStatus.DELETED,
    "Attached": VolumeStatus.IN_USE,
    "Frozen": VolumeStatus.BUSY,
    "ReadyToUpload": VolumeStatus.DELETED,
    "Reserved": VolumeStatus.ERROR,
    "Unattached": VolumeStatus.UNKNOWN,
}
```

Am i correct define this statuses?

# To-Dos

- [x] Implement BaseResources for all Azure compute resources

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

File: plugins/azure/resoto_plugin_azure/resource/compute.py [hunk 1]
```diff
@@ -2499,6 +2530,9 @@ class AzureVirtualMachine(AzureResource):
         "user_data": S("properties", "userData"),
         "virtual_machine_scale_set": S("properties", "virtualMachineScaleSet", "id"),
         "vm_id": S("properties", "vmId"),
+        "instance_cores": S("properties", "hardwareProfile", "vmSizeProperties", "vCPUsAvailable", default=0.0)
+        >> F(lambda x: float(x)),
+        "instance_type": S("type"),
```
Comment (aquamatthias) [inline plugins/azure/resoto_plugin_azure/resource/compute.py:2535]: does this property even exists?

It should be taken from hardwareProfile.vmSize.

File: plugins/azure/resoto_plugin_azure/resource/compute.py [hunk 2]
```diff
@@ -2447,7 +2478,7 @@ class AzureVirtualMachineIdentity:
 
 
 @define(eq=False, slots=False)
-class AzureVirtualMachine(AzureResource):
+class AzureVirtualMachine(AzureResource, BaseInstance):
```
Comment (aquamatthias) [inline plugins/azure/resoto_plugin_azure/resource/compute.py:2481]: I do not see the instance_status field mapped.
It should be derived from the provisioning_state value.

File: plugins/azure/resoto_plugin_azure/resource/compute.py [hunk 3]
```diff
@@ -3040,7 +3074,7 @@ class AzureVirtualMachineScaleSet(AzureResource):
 
 
 @define(eq=False, slots=False)
-class AzureVirtualMachineSize(AzureResource):
+class AzureVirtualMachineSize(AzureResource, BaseInstanceType):
```
Comment (aquamatthias) [inline plugins/azure/resoto_plugin_azure/resource/compute.py:3077]: `instance_type` is not mapped and is the VM size name.

**Review by aquamatthias — COMMENTED** [review]
Thanks. I left a couple of comments.

File: plugins/azure/test/compute_test.py [hunk 4]
```diff
@@ -75,7 +75,15 @@ def test_ssh_key(builder: GraphBuilder) -> None:
...
-    assert len(collected) == 4
+    assert len(collected) == 2
+
+
+def test_virtual_machine_resources(builder: GraphBuilder) -> None:
+    collected = roundtrip_check(AzureVirtualMachine, builder)[0]
+    assert collected.instance_cores == 0.0
+    assert collected.instance_memory == 0.0
+    assert collected.instance_type == "Standard_A0"
+    assert collected.instance_status == InstanceStatus.RUNNING
```
Comment (1101-1) [inline plugins/azure/test/compute_test.py:86]: Do i need to create test that assert more collected resources?

File: plugins/azure/resoto_plugin_azure/resource/compute.py [hunk 5]
```diff
@@ -2446,8 +2478,18 @@ class AzureVirtualMachineIdentity:
     user_assigned_identities: Optional[Dict[str, AzurePrincipalidClientid]] = field(default=None, metadata={'description': 'The list of user identities associated with the virtual machine. The user identity dictionary key references will be arm resource ids in the form: /subscriptions/{subscriptionid}/resourcegroups/{resourcegroupname}/providers/microsoft. Managedidentity/userassignedidentities/{identityname}.'})  # fmt: skip
 
 
+InstanceStatusMapping = {
+    "Creating": InstanceStatus.BUSY,
+    "Updating": InstanceStatus.BUSY,
+    "Succeeded": InstanceStatus.RUNNING,
+    "Failed": InstanceStatus.STOPPED,
+    "Deleting": InstanceStatus.BUSY,
+    "Migrating": InstanceStatus.BUSY,
```
Comment (aquamatthias) [inline plugins/azure/resoto_plugin_azure/resource/compute.py:2487]: The API itself only defines a string and sadly no enumeration.
According to ChatGPT, those are the possible states:
```
Creating: Indicates that the VM is currently being created.
Updating: Indicates that an existing VM is being updated.
Succeeded: Indicates that the last create, update, or delete operation was successful.
Failed: Indicates that the last create, update, or delete operation failed and provides an error message for debugging.
Deleting: Indicates that the VM is in the process of being removed.
Migrating: Indicates that the VM is being moved to a different host server, resource group, or subscription.
Deallocating: Indicates that the VM is being deallocated, which releases compute resources.
Deallocated: Indicates that the VM is deallocated and not consuming compute resources.
Starting: Indicates that the VM is in the process of being started.
Running: Indicates that the VM is currently running.
Stopping: Indicates that the VM is in the process of being stopped.
Stopped: Indicates that the VM is stopped, but still allocated and incurring compute charges.
```

File: plugins/azure/test/compute_test.py [hunk 6]
```diff
@@ -78,6 +78,14 @@ def test_virtual_machine(builder: GraphBuilder) -> None:
     assert len(collected) == 2
 
 
+def test_virtual_machine_resources(builder: GraphBuilder) -> None:
+    collected = roundtrip_check(AzureVirtualMachine, builder)[0]
+    assert collected.instance_cores == 0.0
+    assert collected.instance_memory == 0.0
```
Comment (aquamatthias) [inline plugins/azure/test/compute_test.py:84]: you are not mapping those fields - no need to test them.

File: plugins/azure/test/compute_test.py (hunk 4 above)
Comment (aquamatthias) [reply plugins/azure/test/compute_test.py:86]: not for virtual machine, but disk, snapshot, vmSize, etc.

File: plugins/azure/resoto_plugin_azure/resource/compute.py [hunk 7]
```diff
@@ -582,8 +590,20 @@ class AzureDiskSecurityProfile:
     security_type: Optional[str] = field(default=None, metadata={'description': 'Specifies the securitytype of the vm. Applicable for os disks only.'})  # fmt: skip
 
 
+VolumeStatusMapping = {
+    "ActiveSAS": VolumeStatus.IN_USE,
+    "ActiveSASFrozen": VolumeStatus.BUSY,
+    "ActiveUpload": VolumeStatus.DELETED,
+    "Attached": VolumeStatus.IN_USE,
+    "Frozen": VolumeStatus.BUSY,
+    "ReadyToUpload": VolumeStatus.DELETED,
```
[the remark this answers is not in the document]
Comment (aquamatthias) [reply plugins/azure/resoto_plugin_azure/resource/compute.py:600]: Why should this be marked as deleted?

File: plugins/azure/resoto_plugin_azure/resource/compute.py [hunk 8]
```diff
@@ -2499,6 +2541,8 @@ class AzureVirtualMachine(AzureResource):
         "user_data": S("properties", "userData"),
         "virtual_machine_scale_set": S("properties", "virtualMachineScaleSet", "id"),
         "vm_id": S("properties", "vmId"),
+        "instance_type": S("properties", "hardwareProfile", "vmSize"),
+        "instance_status": S("properties", "provisioningState") >> MapEnum(InstanceStatusMapping, default=InstanceStatus.UNKNOWN),
```
Comment (aquamatthias) [inline plugins/azure/resoto_plugin_azure/resource/compute.py:2545]: Please map `instance_type`

File: plugins/azure/resoto_plugin_azure/resource/compute.py [hunk 9]
```diff
@@ -3063,6 +3107,8 @@ class AzureVirtualMachineSize(AzureResource):
         "number_of_cores": S("numberOfCores"),
         "os_disk_size_in_mb": S("osDiskSizeInMB"),
         "resource_disk_size_in_mb": S("resourceDiskSizeInMB"),
+        "instance_cores": S("numberOfCores"),
+        "instance_memory": S("memoryInMB") >> F(lambda x: int(x) / 1024),
```
Comment (aquamatthias) [inline plugins/azure/resoto_plugin_azure/resource/compute.py:3111]: Please map `instance_type`

**Review by aquamatthias — COMMENTED** [review]
Almost there. I added some comments. PTAL.

File: plugins/azure/test/compute_test.py (hunk 4 above)
Comment (1101-1) [reply plugins/azure/test/compute_test.py:94]: can't add AzureSnapshot resource test, because it collect an empty list:

```
---------------------------------------------------- Captured stdout call -----------------------------------------------------
[]
```

Comment (aquamatthias) [reply plugins/azure/test/compute_test.py:94]: You can. The test setup uses static files to capture an example result.
For snapshots it is [this file](https://github.com/someengineering/resoto/blob/main/plugins/azure/test/files/compute/snapshots.json), which is empty.

You can copy & paste an example from the [Azure REST docs for snapshots](https://learn.microsoft.com/en-us/rest/api/compute/snapshots/list?tabs=HTTP): copy the sample response.

You should see two snapshots.

File: plugins/azure/resoto_plugin_azure/resource/compute.py (hunk 5 above)
Comment (aquamatthias) [reply plugins/azure/resoto_plugin_azure/resource/compute.py:2487]: You marked it as resolved - but I do not see the values mapped?

Comment (1101-1) [reply plugins/azure/resoto_plugin_azure/resource/compute.py:2487]: I found information only about 6 [providing states](https://learn.microsoft.com/en-us/azure/virtual-machines/states-billing).

<div class="heading-wrapper" data-heading-level="h2"><h2 id="provisioning-states" class="heading-anchor">Provisioning states</h2></div>
<p>The provisioning state is the status of a user-initiated, 
control-plane operation on the VM. These states are separate from the 
power state of a VM.</p>
<div class="has-inner-focus">

Provisioning state | Description
-- | --
Creating | Virtual machine is being created.
Updating | Virtual machine is updating to the latest model. Some non-model  changes to a virtual machine such as start and restart fall under the  updating state.
Failed | Last operation on the virtual machine resource was unsuccessful.
Succeeded | Last operation on the virtual machine resource was successful.
Deleting | Virtual machine is being deleted.
Migrating | Seen when migrating from Azure Service Manager to Azure Resource Manager.

</div>


According to`States and billing status of Azure Virtual Machines`, i decide to do not add other states into `InstanceStatusMapping`, because it intended to [PowerState](https://learn.microsoft.com/en-us/azure/virtual-machines/states-billing#power-states-and-billing).

**Review by aquamatthias — DISMISSED** [review]
Thanks!

**Review by aquamatthias — APPROVED** [review]
Nice!

**Review by aquamatthias — APPROVED** [review]
